Sources/XCTest/Public/Factory_Methods.swift

```diff
@@ -0,0 +1,110 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//
+//  Factory_Methods.swift
+//  A collection of methods to create XCTestCase.ClassEntries for use in test code.
+//
+
+/// Wrapper function allowing an array of test case methods to fit
+/// the signature required by `XCTMain`
+/// - seealso: `XCTMain`
+public func testCase<T: XCTestCase>(_ entries: [(String, (T) -> () -> Void)]) -> XCTestCase.ClassEntries {
+    return XCTestCase.ClassEntries(testCaseClass: T.self, entries: entries.map {
+        XCTestCase.Entry(name: $0.0, closure: test($0.1))
+    })
+}
+
+/// Wrapper function allowing an array of non-async, throwing test case methods to fit
+/// the signature required by `XCTMain`
+/// - seealso: `XCTMain`
+public func testCase<T: XCTestCase>(_ entries: [(String, (T) -> () throws -> Void)]) -> XCTestCase.ClassEntries {
+    return XCTestCase.ClassEntries(testCaseClass: T.self, entries: entries.map {
+        XCTestCase.Entry(name: $0.0, closure: test($0.1))
+    })
+}
+
+/// Wrapper function for the async, throwing variant of tests.
+/// the signature required by `XCTMain`
+/// - seealso: `XCTMain`
+public func testCase<T: XCTestCase>(_ entries: [(String, (T) -> () async throws -> Void)]) -> XCTestCase.ClassEntries {
+    return XCTestCase.ClassEntries(testCaseClass: T.self, entries: entries.map {
+        XCTestCase.Entry(name: $0.0, closure: test($0.1))
+    })
+}
+
+/// Wrapper function for the non-async, throwing variant of tests.
+/// the signature required by `XCTMain`
+/// - seealso: `XCTMain`
+@available(*, deprecated, message: "Use the variant returning XCTestCase.ClassEntries instead.")
+@_disfavoredOverload
```

Comment:
Will using `@_disfavoredOverload` here cause the new overloads above which accept `async` closures to be preferred implicitly over their non-async counterparts? I wonder if that will break existing behaviors, by interpreting _every_ test closure as `async` and thus invoking them all via `Task {}`

Reply:
I don't know if the compiler would do that. This sounds like a good question for someone on the concurrency team to explain how the overloads would evaluate.

My suspicion is it _wouldn't_ promote something which isn't explicitly `async` to `async`. That sounds very un-Swifty. But I just don't know. ¯\_(ツ)_/¯

Reply:
Stuart and I were just chatting about this - we're thinking a good next step for this would be to add a bunch of new `isMainThread` assertions into existing *non-async* test cases into the suite, in -setUp, -tearDown, the test method body, and teardown blocks, to help make sure that we're maintaining the expected threading behaviors for existing tests. If Stuart's concern here turns out to be a real problem, we expect that those new assertions would be failing. How does that sound?

Reply:
Sounds like a great plan to me! :D